### PR TITLE
JDK-8259283: use new HtmlId and HtmlIds classes

### PR DESCRIPTION
**Pull request by jonathan-gibbons**
Please review this change to centralize the management of HTML ids used by the standard doclet in a single new factory class, `HtmlIds`, which utilizes a new type-safe wrapper class around `String` called `HtmlId`.

The new classes are used both when declaring ids (e.g. `HtmlTree.setId`) and when referencing them (e.g. `Links.createLink`).  The enum `SectionName`, which declared a set of standard ids, goes away, as do methods in `Links` and other places to create "anchors".

This is a simple refactoring: there is no intentional change of functionality, and no tests are affected or need to be updated. However, the changes do highlight the inconsistency of the use of ids: it would be good to rationalize these in separate, targeted follow-up work.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259283](https://bugs.openjdk.java.net/browse/JDK-8259283): use new HtmlId and HtmlIds classes

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.java.net/census#hannesw) (@hns - **Reviewer**) ⚠️ Review applies to 2ddcf132fc0702cef7e86d7cc02266397f1cf51f

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1951/head:pull/1951`
`$ git checkout pull/1951`

### CODE REVIEW

**Review by hns — APPROVED** [review]
Very nice work, Jon! +1

A few inline comments for an unused field and method and a couple of questions.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Links.java [hunk 1]
```diff
@@ -265,40 +225,14 @@ public Content createLink(DocLink link, Content label, boolean isExternal) {
...
-     *
-     * @param name the string that needs to be converted to a valid HTML id
-     * @return a valid HTML name
+     * @param link       the details for the link
+     * @param label      the content for the link
+     * @return a content tree for the link
      */
-    public String getName(String name) {
-        return name.replaceAll("\\s+", "");
+    public Content createExternalLink(DocLink link, Content label) {
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Links.java:234]: Ah, good catch!  Will fix.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Table.java [hunk 2]
```diff
@@ -299,16 +298,21 @@ public void addRow(Element element, List<Content> contents) {
             int rowIndex = bodyRows.size();
             rowStyle = stripedStyles.get(rowIndex % 2);
         }
-        Set<String> tabClasses = new HashSet<>();
+        Set<String> tabClasses = new HashSet<>(); // !! would be better as a List
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/Table.java:301]: Not yet. I have a list of a number of minor cleanups to do.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlIndexBuilder.java [hunk 3]
```diff
@@ -71,9 +72,10 @@
     HtmlIndexBuilder(HtmlConfiguration configuration) {
         super(configuration, configuration.getOptions().noDeprecated());
         this.configuration = configuration;
-        links = new Links(DocPath.empty, configuration.utils);
+        links = new Links(DocPath.empty);
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlIndexBuilder.java:75]: OK, will check.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlId.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+package jdk.javadoc.internal.doclets.formats.html.markup;
+
+/**
+ * A type-safe wrapper around a {@code String}, for use as an "id"
+ * in {@code HtmlTree} objects.
+ *
+ * @see HtmlTree#setId(HtmlId)
+ */
+public interface HtmlId {
```
[the remark this answers is not in the document]
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlId.java:34]: At one point, I had in mind using an enum class for the set of fixed ids, and so wanted an interface that I could mix in to the declaration of that class.

Eventually, when we have value types, `HtmlId` could be a good candidate for that feature.

**Pull request closed without merging**